### PR TITLE
Java Spring Introduction

### PR DESCRIPTION
**Pull request by artsiompeshko**
NOTES:

- Unit tests were implemented only for one method, just to show how to write them. I believe the goal of the task is not unit tests
- Initialization of storage during the application start part is missed.
- Most logic is naive: no edge case checkings, null pointers, and other production code ready things.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
```
Comment (mikalai-amusin) [inline pom.xml:7]: spring boot?) I suppose for this module you don't need spring boot, just spring core

File: src/main/java/com/apeshko/javaspring/dao/Dao.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.apeshko.javaspring.dao;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface Dao<T> {
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/dao/Dao.java:6]: that's fine, just recommendation, like dao implemented in Spring Data JPA, generic class may be expanded a bit like this (to make it more flexible):
```
public interface Repository<T,ID> {
  T getById(ID id);
    List<T> getAll();
    T save (T t);
    T update(T t);
    boolean delete(ID id);
}
```

File: src/main/java/com/apeshko/javaspring/dao/EventDao.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.apeshko.javaspring.dao;
+
+import java.util.Date;
+import java.util.List;
+
+public interface EventDao<T> extends Dao<T> {
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/dao/EventDao.java:6]: Does EventDao should be really generic class? why not just
```
public interface EventDao extends Dao<Event> {
    List<Event> getEventsByTitle(String title, int pageSize, int pageNum);
    List<Event> getEventsForDay(Date day, int pageSize, int pageNum);
}
```

File: src/main/java/com/apeshko/javaspring/dao/TicketDao.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.apeshko.javaspring.dao;
+
+import com.apeshko.javaspring.model.User;
+
+import java.util.List;
+
+public interface TicketDao<T> extends Dao<T> {
+    List<T> getBookedTicketsByUserId(long userId, int pageSize, int pageNum);
+    List<T> getBookedTicketsByEventId(long eventId, int pageSize, int pageNum);
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/dao/TicketDao.java:9]: better to provide overridden methods to get not paginated data as well, currently we force client to use pagination

File: src/main/java/com/apeshko/javaspring/facade/BookingFacade.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import com.apeshko.javaspring.model.Event;
+import com.apeshko.javaspring.model.Ticket;
+import com.apeshko.javaspring.model.User;
+
+import java.util.Date;
+import java.util.List;
+
+/**
+ * Groups together all operations related to tickets booking.
+ * Created by maksym_govorischev.
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/facade/BookingFacade.java:12]: who is it?)

File: src/main/java/com/apeshko/javaspring/facade/impl/BookingFacadeImpl.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import com.apeshko.javaspring.service.UserService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.List;
+
+@Component
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/facade/impl/BookingFacadeImpl.java]: Seems it more service than component, don't forget about stereotype annotations ;)

File: src/main/java/com/apeshko/javaspring/dao/impl/UserDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import com.apeshko.javaspring.model.User;
+import com.apeshko.javaspring.storage.Storage;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+@Component
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/dao/impl/UserDaoImpl.java]: @Repository

File: src/main/java/com/apeshko/javaspring/facade/impl/BookingFacadeImpl.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    @Override
+    public List<Event> getEventsForDay(Date day, int pageSize, int pageNum) {
+        return eventService.getForDay(day, pageSize, pageNum);
+    }
+
+    @Override
+    public Event createEvent(Event event) {
+        Event createdEvent = eventService.create(event);
+
+        return createdEvent;
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/facade/impl/BookingFacadeImpl.java:52]: we can inline this
`return eventService.create(event);`

File: src/main/java/com/apeshko/javaspring/storage/Storage.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.apeshko.javaspring.storage;
+
+import java.util.List;
+
+public interface Storage {
```
Comment (mikalai-amusin) [inline src/main/java/com/apeshko/javaspring/storage/Storage.java:5]: to avoid class casting, better to create generic interface like with repositories
```
public interface Storage<T> {
    T getByKey(String key);
    List<T> getAllByKeyPrefix(String keyPrefix);
    T save(String key, T value);
    T remove(String key);
}
```

**Comment by artsiompeshko** [comment]
@mikalai-amusin thanks for the review!
